Mathlib/Dynamics/TopologicalEntropy/CoverEntropy.lean

```diff
@@ -0,0 +1,617 @@
+subtypes. Our version directly gives a meaning to the topological entropy of a subsystem, and a
+single theorem (`subset_restriction_entropy` in `TopologicalEntropy.Morphism`) will give the
+equivalence between both versions.
+
+Another choice is to give a meaning to the entropy of `∅` (it must be `-∞` to stay coherent) and to
+keep the possibility for the entropy to be infinite. Hence, the entropy takes values in the extended
+reals `[-∞, +∞]`. The consequence is that we use `ℕ∞`, `ℝ≥0∞` and `EReal` numbers.
+
+## Main definitions
+- `IsDynCoverOf`: property that dynamical balls centered on a subset `s` cover a subset `F`.
+- `coverMincard`: minimal cardinal of a dynamical cover. Takes values in `ℕ∞`.
+- `coverEntropyInfUni`/`coverEntropySupUni`: exponential growth of `coverMincard`. The former is
+defined with a `liminf`, the later with a `limsup`. Take values in `EReal`.
+- `coverEntropyInf`/`coverEntropySup`: supremum of `coverEntropyInfUni`/`coverEntropySupUni` over
+all entourages (or limit as the entourages go to the diagonal). These are Bowen-Dinaburg's
+versions of the topological entropy with covers. Take values in `EReal`.
+
+## Tags
+cover, entropy
+
+## TODO
+The most painful part of many manipulations involving topological entropy is going from
+`coverMincard` to `coverEntropyInfUni`/`coverEntropySupUni`. It involves a logarithm, a division, a
+`liminf`/`limsup`, and multiple coercions. The best thing to do would be to write a file on
+"exponential growth" to make a clean pathway from estimates on `coverMincard` to estimates on
+`coverEntropy`/`coverEntropy'`. It would also be useful in other similar contexts, including the
+definition of entropy using nets.
+
+Get versions of the topological entropy on (pseudo-e)metric spaces.
+-/
+
+namespace Dynamics
+
+open Set Uniformity UniformSpace
+
+variable {X : Type*}
+
+/-! ### Dynamical covers -/
+
+/-- Given a subset `F`, an entourage `U` and an integer `n`, a subset `s` is a `(U, n)`-
+dynamical cover of `F` if any orbit of length `n` in `F` is `U`-shadowed by an orbit of length `n`
+of a point in `s`.-/
+def IsDynCoverOf (T : X → X) (F : Set X) (U : Set (X × X)) (n : ℕ) (s : Set X) : Prop :=
+  F ⊆ ⋃ x ∈ s, ball x (dynEntourage T U n)
+
+lemma IsDynCoverOf.of_le {T : X → X} {F : Set X} {U : Set (X × X)} {m n : ℕ} (m_n : m ≤ n)
+    {s : Set X} (h : IsDynCoverOf T F U n s) :
+    IsDynCoverOf T F U m s := by
+  exact Subset.trans (c := ⋃ x ∈ s, ball x (dynEntourage T U m)) h
+    (iUnion₂_mono fun x _ ↦ ball_mono (dynEntourage_antitone T U m_n) x)
+
+lemma IsDynCoverOf.of_entourage_subset {T : X → X} {F : Set X} {U V : Set (X × X)} (U_V : U ⊆ V)
+    {n : ℕ} {s : Set X} (h : IsDynCoverOf T F U n s) :
+    IsDynCoverOf T F V n s := by
+  exact Subset.trans (c := ⋃ x ∈ s, ball x (dynEntourage T V n)) h
+    (iUnion₂_mono fun x _ ↦ ball_mono (dynEntourage_monotone T n U_V) x)
+
+@[simp]
+lemma isDynCoverOf_empty {T : X → X} {U : Set (X × X)} {n : ℕ} :
```

Comment:
Better say `IsDynCoverOf T ∅ U n s` for an arbitrary `s`. This will be a better simp lemma, as it will apply more often.

Reply:
Done.